File: GenerateParentheses.py
"""
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

"((()))", "(()())", "(())()", "()(())", "()()()"
"""


class Solution:
    # @param {integer} n
    # @return {string[]}
    def generateParenthesis(self, n):
        self.__res = []

        def __generateParentheses(num_left, num_unmatched_left, cur_str):
            if num_left == n and not num_unmatched_left:
                self.__res.append(cur_str)
                return
            if num_left < n:
                __generateParentheses(num_left + 1, num_unmatched_left + 1, cur_str + '(')
            if num_unmatched_left:
                __generateParentheses(num_left, num_unmatched_left - 1, cur_str + ')')

        __generateParentheses(0, 0, "")
        return self.__res


if __name__ == '__main__':
    print Solution().generateParenthesis(3)

File: Java-By-Tag/src/main/GenerateParentheses.java
package src.main;

/**
 * Created by yhf on 10/10/15.
 */

import java.util.ArrayList;
import java.util.List;

/**
 * Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
 *
 * For example, given n = 3, a solution set is:
 *
 * "((()))", "(()())", "(())()", "()(())", "()()()"
 */

public class GenerateParentheses {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        if (n <= 0) {
            return res;
        }
        generateParenthesis(n, 0, "", res);
        return res;
    }

    private void generateParenthesis(int n, int numLeft, String cur, List<String> res) {
        if (cur.length() == 2 * n) {
            if (numLeft == n) {
                res.add(cur);
            }
            return;
        }
        if (numLeft > cur.length() - numLeft) {
            generateParenthesis(n, numLeft, cur + ')', res);
        }
        generateParenthesis(n, numLeft + 1, cur + '(', res);
    }
}
